aiven/datasource_service_component.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2019 Aiven, Helsinki, Finland. https://aiven.io/
+package aiven
+
+import (
+	"fmt"
+	"github.com/aiven/aiven-go-client"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"strconv"
+)
+
+func datasourceServiceComponent() *schema.Resource {
+	return &schema.Resource{
+		Read: datasourceServiceComponentRead,
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Project name",
+			},
+			"service_name": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Service name",
+			},
+			"component": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Service component name",
+			},
+			"route": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Network access route",
+			},
+			"kafka_authentication_method": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Kafka authentication method. This is a value specific to the 'kafka' service component",
+			},
+			"ssl": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Description: "Whether the endpoint is encrypted or accepts plaintext. By default endpoints are " +
+					"always encrypted and this property is only included for service components they may " +
+					"disable encryption",
+			},
+			"usage": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "DNS usage name",
+				Default:     "primary",
+			},
+			"host": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "DNS name for connecting to the service component",
+			},
+			"port": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Port number for connecting to the service component",
+			},
+		},
+	}
+}
+
+func datasourceServiceComponentRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(*aiven.Client)
+
+	projectName := d.Get("project").(string)
+	serviceName := d.Get("service_name").(string)
+	componentName := d.Get("component").(string)
+	route := d.Get("route").(string)
+	usage := d.Get("usage").(string)
+
+	service, err := client.Services.Get(projectName, serviceName)
+	if err != nil {
+		return fmt.Errorf("service %s/%s not found: %w", projectName, serviceName, err)
+	}
+
+	for _, c := range service.Components {
+		if c.Component == componentName && c.Route == route && c.Usage == usage {
+			// check optional ssl search criteria, if not set by a user match entries without ssl
+			if ssl, ok := d.GetOk("ssl"); ok {
+				if c.Ssl != strconv.FormatBool(ssl.(bool)) {
+					continue
+				}
+			} else {
+				if c.Ssl != "" {
```

Comment:
I would accept both empty and true here, otherwise if we don't currently allow non-SSL connections for specific connection type but later add support for it then this would stop matching the same entry it used to match before the change.